### PR TITLE
Refactor gradient clipping

### PR DESCRIPTION
**Pull request by ringgaard**
I have refactored gradient clipping as suggested in issue #237. This also allows be to consolidate the clipping code so it is implemented in the Optimizer class instead of having to implement this in each of the optimizers.

### CODE REVIEW

File: sling/myelin/learning.cc [hunk 1]
```diff
@@ -107,6 +107,17 @@ void Optimizer::Build(Flow *flow) {
     gradmap[var] = tf.Ref(instance_[func], dvar);
   }
 
+  // Optionally add gradient clipping.
+  if (clipping_threshold_ != 0.0) {
+    auto *threshold = tf.Name(tf.Const(clipping_threshold_), "threshold");
+    for (auto &it : gradmap) {
+      auto &dv = it.second;
+      auto *norm = tf.Norm(dv);
```
Comment (rahul1980) [inline sling/myelin/learning.cc]: FYI: This is different from how Pytorch implements clip_grad_norm_()   (which we use in the trainer):
https://pytorch.org/docs/stable/_modules/torch/nn/utils/clip_grad.html

There, the norm is computed over all parameters together (as if the gradients across all params were concatenated together). 

In Tensorflow too, while we have tf.clip_by_norm(tensor, clip_value) and tf.clip_by_global_norm(multiple_tensors, clip_value), the latter seems to be the more popular choice (akin to pytorch).

Comment (ringgaard) [reply sling/myelin/learning.cc]: I have added support for both local and global gradient clipping. However, on the lextagger test, global clipping seems slightly less stable than local clipping, and the final model has ~0.2% lower accuracy.
